Fix configuration generation

danjujan opened the pull request:
Unittest for https://github.com/se-sic/VaRA/issues/1043

danjujan committed:
Add GetAllConfigurations2 unittest to detect bug

vulder commented:
@ChristianKaltenecker Do you have an idea what is wrong with the solver here?

boehmseb committed:
Do not exclude the first config.

For some reason, this caused other configs to appear multiple times during enumeration.

boehmseb committed:
Avoid unnecessary solver calls.

boehmseb committed:
Use for-each loop to iterate maps.

boehmseb committed:
Temporary fix for side effects of getNextConfiguration()

This caused 2 issues:
- either first config was skipped or getCurrentConfig would always point to the next configuration
- getAllValidConfigurations or getNumberValidConfigurations produce incorrect results if called after any call to getNextConfiguration()
The last point is an issue with the Solver API and this is a quick-fix for the Z3Solver implementation that returns an error in that case.

vulder commented on lib/Solver/Z3Solver.cpp:
```diff
@@ -177,35 +177,44 @@ Result<SolverErrorCode, bool> Z3Solver::hasValidConfigurations() {
 Z3Solver::getNextConfiguration() {
-  // Add previous configuration as a constraint
-  excludeCurrentConfiguration();
+  if (!Dirty) {
+    Dirty = true;
+  } else {
+    excludeCurrentConfiguration();
+  }
```
```suggestion
  if (Dirty) {
    excludeCurrentConfiguration();
  } else {
    Dirty = true;
  }
```

vulder commented on lib/Solver/Z3Solver.cpp:
```diff
@@ -258,11 +266,10 @@ Z3Solver::getCurrentConfiguration() {
-  for (auto Iterator = OptionToVariableMapping.begin();
-       Iterator != OptionToVariableMapping.end(); Iterator++) {
-    const z3::expr OptionExpr = *Iterator->getValue();
+  for (auto const &Entry : OptionToVariableMapping) {
```
```suggestion
  for (const auto &Entry : OptionToVariableMapping) {
```

vulder commented on lib/Solver/Z3Solver.cpp:
```diff
@@ -232,18 +241,17 @@ Result<SolverErrorCode> Z3Solver::excludeCurrentConfiguration() {
-  for (auto Iterator = OptionToVariableMapping.begin();
-       Iterator != OptionToVariableMapping.end(); Iterator++) {
-    const z3::expr OptionExpr = *Iterator->getValue();
+  for (auto const &Entry : OptionToVariableMapping) {
```
```suggestion
  for (const auto &Entry : OptionToVariableMapping) {
```

boehmseb committed:
Apply suggested changes.

boehmseb committed:
Quick-fix to make configuration enumeration work.

This implementation is quite hacky and requires a redesign of the Solver API to be fixed properly.

boehmseb commented:
This issue is more complicated than assumed. The current state of this PR should be considered a quick-fix. Do not merge until it is properly resolved.

boehmseb committed:
Change solver interface to better work with internal solver states.

That means that model enumeration is no longer part of the solver interface, but should be done using the ConfigurationFactory and ConfigurationIterator.

boehmseb commented:
This is now ready for review.

vulder commented on include/vara/Solver/ConfigurationFactory.h:
```diff
@@ -17,7 +17,7 @@ class ConfigurationIterable {
   class ConfigurationIterator {
   public:
-    using iterator_category = std::forward_iterator_tag;
+    using iterator_category = std::input_iterator_tag;
```
Why can the config iterator no longer be a forward_iterator? Having it at least forward would probably be good to use it in standard algorithms.

boehmseb replied:
As far as I understood this, a `forward_iterator` allows to iterate multiple times.
That was never possible with the current implementation because incrementing the iterator changes the solver state in a non-reversible way. So it was incorrect to use `forward_iterator` in the first place.

boehmseb replied:
At least the following requirements of a `forward_iterator` are violated:
Given `a` and `b`, dereferenceable iterators of type `It`:
- If `a` and `b` compare equal (`a == b` is contextually convertible to `true`) then either they are both non-dereferenceable or `*a` and `*b` are references bound to the same object.
- `a == b` implies `++a == ++b`

For the first point, it would only be true that `*a` and `*b` describe the same configuration, but they are not the same object (they are  `unique_ptr`s pointing to a `Configuration`). In fact, such an object is created every time the iterator is dereferenced or incremented.

boehmseb replied:
Thinking more about this, I'm not sure if this could also be an issue for the `input_iterator` specification as it says:

If `a` == `b` and `(a, b)` is in the domain of `==` then `*a` is equivalent to `*b`.

This would only be fulfilled if _equivalent_ refers to conceptual equivalence of the objects (i.e., they represent the same ting/configuration). There is currently no equivalence implemented on `Configuration`s.

vulder replied:
Hmm, I see. Then let's keep the inputput tag. PS: I don't think they need to be the same object, so input_iterator should be fine.

boehmseb committed:
Run pre-commit

boehmseb committed:
Fix clang-tidy warnings

vulder committed:
Merge branch 'vara-dev' into fix-configuration-generation

vulder commented:
resolves https://github.com/se-sic/VaRA/issues/1043
